### PR TITLE
KAFKA-10606: Disable auto topic creation for fetch-all-topic-metadata request

### PR DESCRIPTION
**Pull request by Lincong**
There is a bug that causes fetch-all-topic-metadata requests triggering
auto topic creation. Details are described in KAFKA-10606. This is the
simplest way to fix this bug on the broker side.

### CODE REVIEW

**Comment by ijuma** [comment]
Thanks for the PR. Thoughts on how to test this?

**Comment by Lincong** [comment]
@ijuma One way to test it could be to mock objects passed to the `KafkaApis` class. The `MetadataCache` object should be mocked in a way that upon the first invocation on the method `getAllTopics` to simulate the scenario in which the metadata cache got updated "concurrently" with `handleTopicMetadataRequest`.

However, IMO, this change is pretty minor. So, I am wondering how necessary it is to add tests for it?

**Comment by ijuma** [comment]
`KafkaApisTest` has a few tests like that. Could we add one there? The main concern is that we will regress here if we don't have tests.

**Comment by Lincong** [comment]
@ijuma  Will do! Thanks

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (chia7712) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: If we pass ```false``` to getTopicMetadata, it generates ```UNKNOWN_TOPIC_OR_PARTITION``` when the topic is removed, right? If so, does client-side need to handle such error? For example, KafkaAdminClient#listTopics should filter out those "nonexistent" topics (https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java#L1717). Otherwise, users want to get all "existent" topics but response say a_topic is "nonexistent".

**Review by chia7712 — COMMENTED** [review]
@Lincong Nice finding and patch. I left a small question.

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (Lincong) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Hi @chia7712 , thank you for bringing up this issue and it's totally valid! 
In order to preserve its behavior (not including `UNKNOWN_TOPIC_OR_PARTITION ` in the response to fetch all topic metadata request), we implemented the below logic which filters out all entries in the response with `UNKNOWN_TOPIC_OR_PARTITION ` if the metadata request is to get metadata for all topisc.

```
val completeTopicMetadata = (if (metadataRequest.isAllTopics) {
    opicMetadata.filter(_.errorCode() != Errors.UNKNOWN_TOPIC_OR_PARTITION.code())
} else {
    topicMetadata
}) ++ unauthorizedForCreateTopicMetadata ++ unauthorizedForDescribeTopicMetadata
```
We have added a unit test as well. Hope to get some feedback from you soon!

**Comment by chia7712** [comment]
@Lincong Thanks for updating code. Could you fix the conflicting files as well?

**Comment by Lincong** [comment]
> @Lincong Thanks for updating code. Could you fix the conflicting files as well?

Conflicts resolved

**Comment by chia7712** [comment]
@Lincong Could you take a look at those failed tests? I will give a review later :)

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (chia7712) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: Could we do this filter early? For example, how about filtering ```getTopicMetadata``` directly? That makes all fixes be in one place and following authorized ops can skip "unknown partition/topic" as well.

**Review by chia7712 — APPROVED** [review]
@Lincong Thanks for your patch. overall LGTM. Just leave a small suggestion. Please take a look.

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (ijuma) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: This is a good point. We should pass a boolean to `getTopicMetadata` indicating that it's an "allTopics" request and have that method handle everything.

Comment (lmr3796) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Hi @chia7712 @ijuma ,

This is Joseph and I'm @Lincong 's colleague working on this patch with him.  I think this is a good point and just updated the PR accordingly

**Comment by chia7712** [comment]
+1 to last commit. Will merge it to trunk tomorrow if no object.

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala
Comment (chia7712) [inline core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: Could we remove this ```@Before``` code? Changing modifier of ```metadataCache``` from "val" to "var" is enough.

Comment (lmr3796) [reply core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: hey @chia7712 ,

I add this to let each test case to have a clean, uncorrupted `MetadataCache`.  Is there any historical context that we should avoid that cleanup?

Comment (chia7712) [reply core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: It seems to me following code is able to generate a clean, uncorrupted ```MetadataCache``` for each test case.

```
var metadataCache = new MetadataCache(brokerId)
```

Also, it is simpler than ```@Before``` block.

Comment (lmr3796) [reply core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: Hey @chia7712 you're right.  I did some reading and found that I misunderstood JUnit's behavior.  Just updated that part and rebased latest trunk.

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (chia7712) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: @ijuma I will merge this PR tomorrow if there is no objection.

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (ijuma) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: Could we not avoid this altogether instead of doing the `filter` later?

Comment (lmr3796) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: @ijuma I slightly changed it (and rebased latest) so the processing is a branch in the middle about `isFetchAllMetadata`.

It's still a `map` and `filter` structure because I don't want to sacrifice the readability, but it's handled in the middle instead of at the returning statement, which I think is more elegant and readable than the original version.

Let me know what you think!

Comment (lmr3796) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Also, I added the assertion of no UNKNOWN_TOPIC_OR_PARTITION into the unit test

Comment (ijuma) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: I think if you use `flatMap`, you don't need the `filter` and it's still readable. Thoughts?

Comment (lmr3796) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: @ijuma That's a good point.  I've updated it.

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (chia7712) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: Is ```Some/None``` more readable than ```List/Nil```?

**Review by chia7712 — COMMENTED** [review]
@Lincong @lmr3796 Thanks for the updating. one small question is raised. Please take a look.

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (lmr3796) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: @chia7712 Thanks, you're right.  Just updated accordingly and rebased latest trunk

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (ijuma) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: We don't usually include JIRA references in the code unless it's a very complex issue. I think the comment here could be something like:

"A metadata request for all topics should never result in topic auto creation. A topic may be deleted between the creation of `topics` and `topicResponses`, so we make sure to always return `None` for this case."

Comment (lmr3796) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Thanks @ijuma , 

To follow the convention in Kafka, I just updated the comments.  Let me know if you have there are any other thoughts!

**Comment by chia7712** [comment]
@lmr3796 Thanks for your updating. Could you rebase code to include recent big commits to trigger QA again?

**Comment by lmr3796** [comment]
@chia7712 @ijuma I just rebased to latest trunk.  Please take a look

**Comment by chia7712** [comment]
@ijuma I'd like to merge it tomorrow. would be helpful if you could give a final review :)

**Comment by ijuma** [comment]
@chia7712 The code changes seen fine. The comment in KafkaApis still seems a bit convoluted. For example, not returning UnknownTopicOrPartition is not related to backwards compatibility, it doesn't make sense to return that if the given topic wasn't explicitly requested. Similarly, the fact that allowAutoTopicCreation is true for some clients isn't that relevant, the code needs to handle it irrespective of what some clients do or don't do.

I thought my suggested comment in the following was clear and concise, but I'm interested in other opinions:

https://github.com/apache/kafka/pull/9435#discussion_r535321971

**Comment by chia7712** [comment]
> I thought my suggested comment in the following was clear and concise, but I'm interested in other opinions:
> #9435 (comment)

@ijuma Sorry that I neglected that conversation.

@lmr3796 FYI ~

**Comment by lmr3796** [comment]
Hi thanks @ijuma @chia7712 ,

I updated the comments accordingly to make it more concise and rebased latest trunk.
Please let me know if there are any other issues.

**Comment by chia7712** [comment]
> I updated the comments accordingly to make it more concise and rebased latest trunk.
Please let me know if there are any other issues.

@lmr3796 Thanks for updating. waiting for response from @ijuma :)

**Comment by ijuma** [comment]
@chia7712 feel free to merge if it looks good to you. Thanks @lmr3796 for the patience.

**Comment by chia7712** [comment]
@Lincong @lmr3796 Thanks for your efforts!

**Pull request merged**
